go-controller/pkg/crd/clusternetworkconnect/v1/types.go

```diff
@@ -0,0 +1,189 @@
+	//  any of other connect subnets for other ClusterNetworkConnects that might be selecting same networks.
+	//
+	// Does not have a default value for the above reason so
+	// that user takes care in setting non-overlapping subnets.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=2
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="connectSubnets is immutable"
+	// +kubebuilder:validation:XValidation:rule="size(self) != 2 || !isCIDR(self[0].cidr) || !isCIDR(self[1].cidr) || cidr(self[0].cidr).ip().family() != cidr(self[1].cidr).ip().family()", message="When 2 CIDRs are set, they must be from different IP families"
+	ConnectSubnets []ConnectSubnet `json:"connectSubnets"`
+
+	// connectivityEnabled specifies which connectivity types should be enabled for the connected networks.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=2
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="connectivityEnabled cannot contain duplicate values"
+	ConnectivityEnabled []ConnectivityType `json:"connectivityEnabled"`
+}
+
+// +kubebuilder:validation:XValidation:rule="isCIDR(self) && cidr(self) == cidr(self).masked()", message="CIDR must be a valid network address"
+// +kubebuilder:validation:MaxLength=43
+type CIDR string
+
+// +kubebuilder:validation:XValidation:rule="!has(self.networkPrefix) || !isCIDR(self.cidr) || self.networkPrefix > cidr(self.cidr).prefixLength()", message="NetworkPrefix must be smaller than CIDR subnet"
+// +kubebuilder:validation:XValidation:rule="!has(self.networkPrefix) || !isCIDR(self.cidr) || (cidr(self.cidr).ip().family() != 4 || self.networkPrefix < 32)", message="NetworkPrefix must < 32 for ipv4 CIDR"
+type ConnectSubnet struct {
+	// CIDR specifies ConnectSubnet, which is split into smaller subnets for every connected network.
+	// This CIDR should be containing 2*((Number of L3 networks*Max Number of Nodes)+Number of L2 networks) IPs.
+	// Example: cidr= "192.168.0.0/16", networkPrefix= 24 means that you can connect 256 Layer3 networks OR
+	// 255 Layer3 networks and 128 Layer2 networks
+	//
+	// CIDR also restricts the maximum number of networks that can be connected together
+	// based on what CIDR range is picked. So choosing a large enough CIDR for future use cases
+	// is important.
+	// +required
+	CIDR CIDR `json:"cidr"`
+
+	// NetworkPrefix specifies the prefix length for every connected network.
+	// This prefix length should be equal to or longer than the length of the CIDR prefix.
+	//
+	// For example, if the CIDR is 10.0.0.0/8 and the networkPrefix is 24,
+	// then the connect subnet for each connected layer3 network will be 10.0.0.0/24, 10.0.1.0/24, 10.0.2.0/24 etc.
+	// For each connected layer2 network we will allocate a /networkPrefix range
```

Comment:
not for each right? should we say
for layer2 networks we will allocate the next available /networkPrefix range that is then split into /31 or /127 slices for each layer2 network